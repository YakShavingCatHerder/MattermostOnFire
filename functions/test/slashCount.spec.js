'use strict';
/* Imports and Mocks */
let admin = require('firebase-admin');
admin.initializeApp = jest.fn();
const test = require('firebase-functions-test')();

const {
  TEST_MM_INTEGRATION_TOKEN,
  TEST_BASE_URL,
  VALID_COUNT_REQUEST_BODY,
  TEST_ACTIVE_POLL
} = require('./sampleData');

test.mockConfig({
  mattermost: { token: TEST_MM_INTEGRATION_TOKEN },
  functions: { baseurl: TEST_BASE_URL }
});

const utils = require('../utils');
const myFunctions = require('../index');

describe('slashCount', () => {
  let refMock, databaseStub;

  beforeAll(() => {
    refMock = jest.fn(location => {
      return {
        once: jest.fn(eventName => Promise.resolve({val: jest.fn(() => TEST_ACTIVE_POLL)}))
      };
    })
    databaseStub = jest.fn(() => {
      return {
        ref: refMock
      };
    });
    Object.defineProperty(admin, "database", { get: () => databaseStub });
  });

  it('Should check for a missing token', done => {
    let mockRequest = { body: utils.deepCopy(VALID_COUNT_REQUEST_BODY) };
    delete mockRequest.body.context.token;
    const mockResponse = {
      status: code => {
        expect(code).toEqual(401);
        return {
          send: jest.fn(text => {
            expect(text).toMatchSnapshot();
            done();
          })
        };
      }
    };
    myFunctions.slashCount(mockRequest, mockResponse);
  });

  it('Should check for a bad token', done => {
    let mockRequest = { body: utils.deepCopy(VALID_COUNT_REQUEST_BODY) };
    mockRequest.body.context.token = 'invalid token';
    const mockResponse = {
      status: code => {
        expect(code).toEqual(401);
        return {
          send: jest.fn(text => {
            expect(text).toMatchSnapshot();
            done();
          })
        };
      }
    };
    myFunctions.slashCount(mockRequest, mockResponse);
  });

  it('Should check for insufficient permissions', done => {
    let mockRequest = { body: utils.deepCopy(VALID_COUNT_REQUEST_BODY) };
    mockRequest.body.user_id = 'not poll creator';

    const mockResponse = {
      set: jest.fn(),
      status: code => {
        expect(code).toEqual(200);
        return {
          send: jest.fn(responseObject => {
            expect(refMock).toHaveBeenCalled();
            expect(responseObject).toMatchSnapshot();
            done();
          })
        };
      }
    };
    myFunctions.slashCount(mockRequest, mockResponse);
  });

  it('Should perform a successful count', done => {
    let mockRequest = { body: utils.deepCopy(VALID_COUNT_REQUEST_BODY) };
    const mockResponse = {
      set: jest.fn(),
      status: code => {
        expect(code).toEqual(200);
        return {
          send: jest.fn(responseObject => {
            expect(refMock).toHaveBeenCalled();
            expect(responseObject).toMatchSnapshot();
            done();
          })
        };
      }
    };
    myFunctions.slashCount(mockRequest, mockResponse);
  });
});
