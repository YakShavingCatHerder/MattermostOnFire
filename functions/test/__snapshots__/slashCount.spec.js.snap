// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`slashCount Should check for a bad token 1`] = `"Invalid request or missing token"`;

exports[`slashCount Should check for a missing token 1`] = `"Invalid request or missing token"`;

exports[`slashCount Should check for insufficient permissions 1`] = `
Object {
  "ephemeral_text": "Unable to get vote count for poll. Only the poll creator can get the vote count. Please ask @jed.fonner if you need the count.",
}
`;

exports[`slashCount Should perform a successful count 1`] = `
Object {
  "ephemeral_text": "Current summary of poll \\"**Does this thing still work?**\\"

| OPTION | VOTES: 1 | PERCENT |
| :----- | -----: | -----: |
| Yes | 1 | 100% |
| No | 0 | 0% |",
}
`;
